src/pulse-tracing-qvm.lisp

```diff
@@ -0,0 +1,221 @@
+  (check-type qvm pulse-tracing-qvm)
+  (remove-if-not (lambda (f)
+                   (quil::frame-intersects-p f qubits))
+                 (alexandria:hash-table-keys (frame-states qvm))))
+
+(defun frames-on-qubits (qvm qubits)
+  "Return all frames tracked by the pulse tracing QVM which involve exactly the specified QUBITS in the specified order."
+  (check-type qvm pulse-tracing-qvm)
+  (remove-if-not (lambda (f)
+                   (quil::frame-on-p f qubits))
+                 (alexandria:hash-table-keys (frame-states qvm))))
+
+(defmethod transition ((qvm pulse-tracing-qvm) (instr quil:delay-on-frames))
+  (dolist (frame (quil:delay-frames instr))
+    (incf (local-time qvm frame) (quil:delay-duration instr)))
+
+  (incf (pc qvm))
+  qvm)
+
+(defmethod transition ((qvm pulse-tracing-qvm) (instr quil:delay-on-qubits))
+  (let ((frames (frames-on-qubits qvm (quil:delay-qubits instr))))
+    ;; delegate
+    (transition qvm (make-instance 'quil:delay-on-frames
+                                   :frames frames
+                                   :duration (quil:delay-duration instr)))))
+
+(defun synchronize-frame-clocks (qvm frames)
+  "Bring all clocks on the specified FRAMES to their latest time."
+  (let ((latest (apply #'latest-time qvm frames)))
+    (dolist (frame frames)
+      (setf (local-time qvm frame) latest))))
+
+(defmethod transition ((qvm pulse-tracing-qvm) (instr quil:fence))
+  (let ((frames (intersecting-frames qvm (quil:fence-qubits instr))))
+    (synchronize-frame-clocks qvm frames))
+
+  (incf (pc qvm))
+  qvm)
+
+(defmethod transition ((qvm pulse-tracing-qvm) (instr quil:simple-frame-mutation))
+  (let* ((frame (quil:frame-mutation-target-frame instr))
+         (val (quil:constant-value
+               (quil:frame-mutation-value instr)))
+         (fs (frame-state qvm frame)))
+
+    ;; update state
+    (etypecase instr
+      (quil:set-frequency
+       (setf (frame-state-frequency fs) val))
+      (quil:set-phase
+       (setf (frame-state-phase fs) val))
+      (quil:shift-phase
+       (incf (frame-state-phase fs) val))
+      (quil:set-scale
+       (setf (frame-state-scale fs) val)))
+
+    ;; update entry
+    (setf (frame-state qvm frame) fs))
+
+  (incf (pc qvm))
+  qvm)
+
+(defmethod transition ((qvm pulse-tracing-qvm) (instr quil:swap-phase))
+  (with-slots (left-frame right-frame) instr
+    (when (equalp left-frame right-frame)
+      (error "SWAP-PHASE requires distinct frames."))
+    ;; there is an implicit synchronization in SWAP-PHASE
+    (apply #'synchronize-frame-clocks left-frame right-frame)
+    (let ((left-state (frame-state qvm left-frame))
+          (right-state (frame-state qvm right-frame)))
+      (rotatef (frame-state-phase left-state) (frame-state-phase right-state))
+      (setf (frame-state qvm left-frame) left-state
+            (frame-state qvm right-frame) right-state)))
+
+  (incf (pc qvm))
+  qvm)
+
+(defmethod transition ((qvm pulse-tracing-qvm) instr)
```

Comment:
this blanket matching on instr makes me nervous. Isn't there a quil instruction superclass at least? or would any of that truly just be noise?

i at least rec specializing on quil:pulse/quil:{raw-}capture